sdk/search/azure-search-documents/azure/search/documents/_service/_search_service_client.py

```diff
@@ -554,3 +555,114 @@ def create_or_update_skillset(self, name, **kwargs):
+
+            .. literalinclude:: ../samples/sample_data_source_operations.py
+                :start-after: [START create_data_source]
+                :end-before: [END create_data_source]
+                :language: python
+                :dedent: 4
+                :caption: Create a Data Source
+        """
+        kwargs["headers"] = self._merge_client_headers(kwargs.get("headers"))
+        result = self._client.data_sources.create(data_source, **kwargs)
+        return result
+
+    @distributed_trace
+    def create_or_update_datasource(self, data_source, name=None, **kwargs):
+        # type: (str, DataSource, Optional[str], **Any) -> Dict[str, Any]
+        """Creates a new datasource or updates a datasource if it already exists.
+
+        :param name: The name of the datasource to create or update.
+        :type name: str
+        :param data_source: The definition of the datasource to create or update.
+        :type data_source: ~search.models.DataSource
+        :return: The created DataSource
+        :rtype: dict
+        """
+        # TODO: access_condition
+        kwargs["headers"] = self._merge_client_headers(kwargs.get("headers"))
+
+        if not Name:
+            name = data_source.name
+        result = self._client.data_sources.create_or_update(name, data_source, **kwargs)
+        return result
+
+    @distributed_trace
+    def get_datasource(self, name, **kwargs):
+        # type: (str, **Any) -> Dict[str, Any]
+        """Retrieves a datasource definition.
+
+        :param name: The name of the datasource to retrieve.
+        :type name: str
+        :return: The DataSource that is fetched.
+        :rtype: dict
+
+        .. admonition:: Example:
+
+            .. literalinclude:: ../samples/sample_data_source_operations.py
+                :start-after: [START get_data_source]
+                :end-before: [END get_data_source]
+                :language: python
+                :dedent: 4
+                :caption: Retrieve a DataSource
+        """
+        kwargs["headers"] = self._merge_client_headers(kwargs.get("headers"))
+        result = self._client.data_sources.get(name, **kwargs)
+        return result
+
+    @distributed_trace
+    def list_datasources(self, **kwargs):
+        # type: (**Any) -> Sequence[DataSource]
+        """Lists all datasources available for a search service.
+
+        :return: List of all the data sources.
+        :rtype: `list[dict]`
+
+        .. admonition:: Example:
+
+            .. literalinclude:: ../samples/sample_data_source_operations.py
+                :start-after: [START list_data_source]
+                :end-before: [END list_data_source]
+                :language: python
+                :dedent: 4
+                :caption: List all the DataSources
+        """
+        kwargs["headers"] = self._merge_client_headers(kwargs.get("headers"))
+        result = self._client.data_sources.list(**kwargs)
+        return result.data_sources
```

Comment:
per @johanste This will either need to return a `PagedItem` or be renamed to `get_datasources` (same comment for async version) @brjohnstmsft has indicated that the only one that realistically *might* be paged on the service side is the indexes list.

Reply:
Could you confirm if there is an upper limit for the number of data sources?

Reply:
I will implement paging in a separate issue - that way this wont be blocked for indexers

https://github.com/Azure/azure-sdk-for-python/issues/11025

Reply:
So with #11025 closed, this should be renamed to `get_datasources` I think?